### PR TITLE
PIMS-1250 Add config and secrets to support health checks and rocketc…

### PR DESCRIPTION
**Pull request by devinleighsmith**
…hat notifications.

Please see comments, as there are some things to think about with this functionality.
This is using this package: https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks

### CODE REVIEW

File: backend/api/Controllers/LookupController.cs [hunk 1]
```diff
@@ -57,7 +57,7 @@ public IActionResult GetAgencies()
         /// Get all of the role code values
         /// </summary>
         /// <returns></returns>
-        [HttpGet("agencies")]
```
Comment (devinleighsmith) [inline backend/api/Controllers/LookupController.cs]: this is an unrelated bugfix.

File: backend/api/Startup.cs [hunk 2]
```diff
@@ -115,15 +118,15 @@ public void ConfigureServices(IServiceCollection services)
                 options.AddPolicy("Administrator", policy => policy.Requirements.Add(new RealmAccessRoleRequirement("administrator")));
```
Comment (devinleighsmith) [inline backend/api/Startup.cs]: I recommend hiding whitespace changes for this file.

File: backend/api/Startup.cs [hunk 3]
```diff
@@ -115,15 +118,15 @@ public void ConfigureServices(IServiceCollection services)
...
             });
 
+            var cs = Configuration.GetConnectionString("PIMS");
+            var builder = new SqlConnectionStringBuilder(cs);
+            var pwd = Configuration["DB_PASSWORD"];
+            if (!String.IsNullOrEmpty(pwd))
+            {
+                builder.Password = pwd;
+            }
             services.AddDbContext<PimsContext>(options =>
```
Comment (devinleighsmith) [inline backend/api/Startup.cs]: Moving the logic to get the connection builder string out of  the if

File: backend/api/Startup.cs [hunk 4]
```diff
@@ -138,6 +141,20 @@ public void ConfigureServices(IServiceCollection services)
...
 
             services.AddAutoMapper(typeof(Startup));
+
+            services.AddHealthChecks()
+                .AddCheck("liveliness", () => HealthCheckResult.Healthy())
+                .AddSqlServer(builder.ConnectionString, tags: new[] { "services" });
+            //TODO: Add a health check for keycloak connectivity.
+            services.AddHealthChecksUI(setupSettings: setup =>
+            {
+                setup.AddHealthCheckEndpoint("liveliness", "http://localhost/live");
```
Comment (devinleighsmith) [inline backend/api/Startup.cs]: While these are hardcoded, I think that is OK since our pims api server would always also be hosting the health check ui interface.

File: backend/api/Startup.cs [hunk 5]
```diff
@@ -138,6 +141,20 @@ public void ConfigureServices(IServiceCollection services)
             services.AddScoped<IKeycloakRequestClient, KeycloakRequestClient>();
 
             services.AddAutoMapper(typeof(Startup));
+
+            services.AddHealthChecks()
+                .AddCheck("liveliness", () => HealthCheckResult.Healthy())
```
Comment (devinleighsmith) [inline backend/api/Startup.cs]: this is just a simple function that always returns true.

File: backend/api/Startup.cs [hunk 6]
```diff
@@ -138,6 +141,20 @@ public void ConfigureServices(IServiceCollection services)
...
+                .AddSqlServer(builder.ConnectionString, tags: new[] { "services" });
```
Comment (devinleighsmith) [inline backend/api/Startup.cs]: this executes select 1 against the database in the connection string.

File: backend/api/Startup.cs [hunk 7]
```diff
@@ -138,6 +141,20 @@ public void ConfigureServices(IServiceCollection services)
...
+            //TODO: Add a health check for keycloak connectivity.
+            services.AddHealthChecksUI(setupSettings: setup =>
```
Comment (devinleighsmith) [inline backend/api/Startup.cs]: This adds an interface that provides the following functionality:
1) View the status of health checks within the application.
2) Send notifications in the event that one or more health checks fail, see below for more details.

File: backend/api/Startup.cs [hunk 8]
```diff
@@ -138,6 +141,20 @@ public void ConfigureServices(IServiceCollection services)
...
+            {
+                setup.AddHealthCheckEndpoint("liveliness", "http://localhost/live");
+                setup.AddHealthCheckEndpoint("readiness", "http://localhost/ready");
+                setup.AddWebhookNotification("rocketchat",
+                    uri: Configuration["ROCKETCHAT_HEALTH_HOOK"],
```
Comment (devinleighsmith) [inline backend/api/Startup.cs]: These are stored as secrets. The two values below could also be hardcoded in theory, but I like knowing which env the healthchecks are failing in.

File: backend/api/Startup.cs (hunk 7 above)
Comment (devinleighsmith) [reply backend/api/Startup.cs]: Our openshift pipeline should notify us if the health checks fail during the deployment process, but this provides the extra functionality of notifying us if the health checks fail during normal execution.

**Comment by devinleighsmith** [comment]
If this all works in Openshift I'll add config to our deployment to actually consume the new healthcheck endpoints.

File: backend/api/Startup.cs [hunk 9]
```diff
@@ -115,15 +118,15 @@ public void ConfigureServices(IServiceCollection services)
                 options.AddPolicy("Administrator", policy => policy.Requirements.Add(new RealmAccessRoleRequirement("administrator")));
             });
 
+            var cs = Configuration.GetConnectionString("PIMS");
```
Comment (Fosol) [inline backend/api/Startup.cs]: I wonder if the connection string is missing whether all this fails in a nice and polite way?

File: backend/api/Startup.cs (hunk 5 above)
Comment (Fosol) [reply backend/api/Startup.cs]: Is this an endpoint `/api/liveliness`?

File: backend/api/Startup.cs (hunk 6 above)
Comment (Fosol) [reply backend/api/Startup.cs]: What happens if it fails?

File: backend/api/Startup.cs (hunk 7 above)
Comment (Fosol) [reply backend/api/Startup.cs]: I'm very curious to see all this working.

File: backend/api/Startup.cs (hunk 4 above)
Comment (Fosol) [reply backend/api/Startup.cs]: 1) Make sure all routes are `/api/...`
2) I don't understand the `http://localhost/live` part.  I would think this needs to be configurable, handle https, host and port.

File: backend/api/Startup.cs (hunk 8 above)
Comment (Fosol) [reply backend/api/Startup.cs]: When does this get called/triggered?

File: backend/api/Startup.cs [hunk 10]
```diff
@@ -166,10 +183,20 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             app.UseAuthorization();
 
             app.UseMiddleware(typeof(LogRequestMiddleware));
-
+            app.UseHealthChecks("/live", new HealthCheckOptions
```
Comment (Fosol) [inline backend/api/Startup.cs]: Is this now the actual route `/live` rather than `/liveliness`?  I don't understand the differences.  Also routes for the api need to be consistent `/api/...` or we'll have two different paths depending on the caller path.

File: backend/api/Startup.cs (hunk 4 above)
Comment (devinleighsmith) [reply backend/api/Startup.cs]: /live is the endpoint that I've configured below.
/live is hosted on our server. however, healthchecksui doesn't know that. HealthChecksUI can be configured to run locally on the same server as the health checks, or remotely. in this case we are running locally, thus the http://localhost. According to the docs just having just having /live works in some cases, but they still have some issues (it didn't work for me within docker).

Https should work, but also shouldn't matter since the connection isn't crossing a network.
Same with being configurable, that shouldn't be necessary unless we plan on having HealthChecksUI not run on the same server. If that is the case then we could create a separate container, and then there are completely different connection parameters that can be used to detect health endpoints on kubernetes.

File: backend/api/Startup.cs (hunk 5 above)
Comment (devinleighsmith) [reply backend/api/Startup.cs]: currently just liveliness, but I can add the /api

File: backend/api/Startup.cs (hunk 6 above)
Comment (devinleighsmith) [reply backend/api/Startup.cs]: the health check returns unhealthy.

File: backend/api/Startup.cs (hunk 8 above)
Comment (devinleighsmith) [reply backend/api/Startup.cs]: Triggered every 10 minutes when a health check is failing (configurable) or once when a failing health check stops failing.

File: backend/api/Startup.cs (hunk 9 above)
Comment (devinleighsmith) [reply backend/api/Startup.cs]: I'll test this.

File: backend/api/Startup.cs (hunk 10 above)
Comment (devinleighsmith) [reply backend/api/Startup.cs]: liveliness is not the route, it is just the name of the health check.
Step 1. Create health checks and add names/tags to them as necesssary.
Step 2. Map all health checks to endpoints based on the identifiers created in Step 1.

File: backend/api/Startup.cs (hunk 4 above)
Comment (devinleighsmith) [reply backend/api/Startup.cs]: I've looked into this a bit more, while I don't want to get hung up on the details too much for this SPIKE, we could configure most of the health checks ui within appsettings.json, for example:

{
  "HealthChecksUI": {
    "HealthChecks": [
      {
        "Name": "HTTP-Api-Basic",
        "Uri": "http://localhost:6457/healthz"
      }
    ],
    "Webhooks": [
      {
        "Name": "",
        "Uri": "",
        "Payload": "",
        "RestoredPayload": ""
      }
    ],
    "EvaluationTimeInSeconds": 10,
    "MinimumSecondsBetweenFailureNotifications": 60
  }
}
we would still need to inject the value of the webhook secret here, but this might be more what you are looking for.

File: backend/api/Startup.cs (hunk 9 above)
Comment (devinleighsmith) [reply backend/api/Startup.cs]: Yes, it seems to handle that in a reasonable manner:

image

File: backend/api/Startup.cs (hunk 7 above)
Comment (asanchezr) [reply backend/api/Startup.cs]: I also like that we get a nice UI but I'd suggest if you could check it out a bit so it can't get abused to spam our server(s)
